File: arduino/TinyPacks/TinyPostman.cpp
//  TinyPostman - Copyright (c) 2012 Francisco Castro <http://fran.cc>
//
//  Permission is hereby granted, free of charge, to any person obtaining a
//  copy of this software and associated documentation files (the "Software"),
//  to deal in the Software without restriction, including without limitation
//  the rights to use, copy, modify, merge, publish, distribute, sublicense,
//  and/or sell copies of the Software, and to permit persons to whom the
//  Software is furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
// 
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
//  THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
//  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
//  DEALINGS IN THE SOFTWARE.


#include "TinyPostman.h"

// Postman

Postman::Postman()
{
    registered_resources = 0;
}

bool Postman::registerResource(const char * path, Resource &resource)
{
    if(registered_resources < TPM_MAX_RESOURCES) {
        resources[registered_resources].path = path;
        resources[registered_resources].resource = &resource;
        registered_resources += 1;
        return true;
    }
    else
        return false;
}

tp_length_t Postman::handlePack(uint8_t * buffer, tp_length_t length, tp_length_t max_length) 
{
    uint8_t method;
    uint8_t response;
    uint8_t i;
    
    request.reader.setBuffer(buffer, length);
    request.writer.setBuffer(buffer, max_length - 2);
    request.writer.setOffset(2);

    if(!request.reader.next() || !request.reader.isInteger() || !(method = request.reader.getInteger()))
        response = TPM_400_Bad_Request;
    else if(!request.reader.next() || !request.writer.setOffset(request.reader.elementStart() + request.reader.elementLength() - buffer))
        response = TPM_400_Bad_Request;        
    else if(!request.reader.next() || !request.reader.isString() || request.reader.getString(request.path, TPM_MAX_PATH_LENGTH) == TP_INVALID_LENGTH)
        response = TPM_400_Bad_Request;
    else if(method == TPM_GET && !request.path[0]) {
        request.writer.openList();
        for(i = 0; i != registered_resources; i++)
            request.writer.putString(resources[i].path);
        request.writer.close();
        response = TPM_205_Content;
    }
    else {
        for(i = 0; i != registered_resources; i++) {
            if(!strcmp(resources[i].path, request.path)) {
                if(method == TPM_GET) 
                    response = resources[i].resource->get(request);
                else if(method == TPM_POST) 
                    response = resources[i].resource->post(request);
                else if(method == TPM_PUT) 
                    response = resources[i].resource->put(request);
                else if(method == TPM_DELETE)
                    response = resources[i].resource->del(request);
                else
                    response = TPM_400_Bad_Request;
                break;
            }
        }
        if(i == registered_resources)
            response = TPM_404_Not_Found;
    }
    buffer[0] = TP_INTEGER | 1;
    buffer[1] = response;
    return request.writer.getOffset();
}


// Framer

Framer::Framer(uint8_t * pack_buffer, tp_length_t pack_max_length) {
    buffer = pack_buffer;
    max_length = pack_max_length;
    state = TPM_RECEIVING;
    crc = crc1 = crc2 = 0;
    length = 0;
    index = 0;
    escape = false;
}

bool Framer::putReceivedByte(uint8_t value) {
    bool valid_frame = false;
    if(state == TPM_RECEIVING) {
        if(value == 0x7E) {
            valid_frame = index > 2 && crc2 == (buffer[index - 2] << 8 | buffer[index - 1]);
            length = index - 2;
            setState(TPM_RECEIVING);
        }
        else if(value == 0x7D)
            escape = true;
        else {
            value = escape ? value ^ 0x20 : value;
            escape = false;
            crc2 = crc1;
            crc1 = crc;
            crc16(crc, value);
            if(index < max_length)
                buffer[index++] = value;
        }
    }
    return valid_frame;
}

uint8_t Framer::getByteToSend() {
    if(state == TPM_SENDING) {
        if(index == length + 2) {
                setState(TPM_RECEIVING);
                return 0x7E;
        }
        else if(index == length) {
                buffer[index] = crc >> 8;
                buffer[index + 1] = crc & 0xFF;
        }
        
        if(escape) {
            escape = false;
            crc16(crc, buffer[index]);
            return buffer[index++] ^ 0x20;
        }
        else if(buffer[index] == 0x7E || buffer[index] == 0x7D) {
            escape = true;
            return 0x7D;
        }
        else {
            crc16(crc, buffer[index]);
            return buffer[index++];
        }
    }
    else
        return 0x7E;
};

File: python/tinypostman/tinypostman.py
#!/usr/bin/python
#
#  TinyPostman - Copyright (c) 2012 Francisco Castro <http://fran.cc>
#
#  Permission is hereby granted, free of charge, to any person obtaining a
#  copy of this software and associated documentation files (the "Software"),
#  to deal in the Software without restriction, including without limitation
#  the rights to use, copy, modify, merge, publish, distribute, sublicense,
#  and/or sell copies of the Software, and to permit persons to whom the
#  Software is furnished to do so, subject to the following conditions:
#
#  The above copyright notice and this permission notice shall be included in
#  all copies or substantial portions of the Software.
# 
#  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
#  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
#  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
#  THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
#  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
#  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
#  DEALINGS IN THE SOFTWARE.

import time
import crc16
import struct
import serial
import tinypacks

TPM_GET    = 0x01
TPM_POST   = 0x02
TPM_PUT    = 0x03
TPM_DELETE = 0x04

TPM_201_Created            = 0x21
TPM_202_Deleted            = 0x22
TPM_204_Changed            = 0x24
TPM_205_Content            = 0x25
TPM_400_Bad_Request        = 0x40
TPM_401_Unauthorized       = 0x41
TPM_403_Forbidden          = 0x43
TPM_404_Not_Found          = 0x44
TPM_405_Method_Not_Allowed = 0x45
TPM_413_Request_Entity_Too_Large     = 0x4D

TPM_RESPONSE_TEXT = {
    0x21: "201 Created",
    0x22: "202 Deleted",
    0x24: "204 Changed",
    0x25: "205 Content",
    0x40: "400 Bad Request",
    0x41: "401 Unauthorized",
    0x43: "403 Forbidden",
    0x44: "404 Not Found",
    0x45: "405 Method Not Allowed",
    0x4D: "413 Request Entity Too Large",
}

class PostmanError(IOError):
    pass
    
class Postman:
    def __init__(self, device, timeout=4):
        self.debug = False
        self.token = 0
        try:
            self.fd = serial.Serial(device, 9600, stopbits=serial.STOPBITS_ONE, parity=serial.PARITY_NONE, timeout=timeout)
        except IOError as err:
            raise PostmanError("Failed to open port: %s" % err)

    def send(self, method, token, path, has_payload=False, payload=None):
        if has_payload:
            frame = "%s%s%s%s" % (tinypacks.pack(method), tinypacks.pack(token), tinypacks.pack(path), tinypacks.pack(payload))
        else:
            frame = "%s%s%s" % (tinypacks.pack(method), tinypacks.pack(token), tinypacks.pack(path))
        frame +=  struct.pack(">H", crc16.crc16str(frame))

        if self.debug:
            print "TX:",
            for c in frame:
                print "%X" % ord(c),
            print "[%i]" % len(frame)

        try:
            self.fd.write("\x7E")
            for byte in frame:
                if byte == "\x7E":
                    self.fd.write("\x7D\x5E")
                elif byte == "\x7D":
                    self.fd.write("\x7D\x5D")
                else:
                    self.fd.write(byte)
            self.fd.write("\x7E")
        except IOError as err:
            raise PostmanError("Frame sending failed: %s" % err)

    def send_null(self):
        try:
            self.fd.write("\x7E")
        except  IOError as err:
            raise PostmanError("Frame sending failed: %s" % err)

    def receive(self):
        frame = ""
        byte = ""
        try:
            while (byte != "\x7E" or frame == ""):
                byte = self.fd.read(1)
                if not byte:
                    raise PostmanError("Frame receiving timed out.")
                if byte == "\x7D":
                    frame += chr(ord(self.fd.read(1))|32)
                elif byte != "\x7E":
                    frame += byte
        except PostmanError:
            raise
        except  IOError as err:
            raise PostmanError("Frame receiving failed: %s" % err)

        if self.debug:
            print "RX:",
            for c in frame:
                print "%X" % ord(c),
            print "[%i]" % len(frame)

        if len(frame) >= 4 and crc16.crc16str(frame[:-2]) == struct.unpack(">H", frame[-2:])[0]:
            frame = frame[:-2]
            response, frame = tinypacks.unpack(frame)
            if frame:
                token, frame = tinypacks.unpack(frame)
                if frame:
                    payload, frame = tinypacks.unpack(frame)
                    return [response, token, payload]
                else:
                    return [response, token]
            else:
                return [response]
        else:
            raise PostmanError("Frame receiving failed, bad CRC %s != %s " % (hex(crc16.crc16str(frame[:-2])), hex(struct.unpack(">H", frame[-2:])[0])))

    def get(self, path, query=None):
        self.send(TPM_GET, self.token, path, query is not None, query)
        response = self.receive()
        if len(response) > 1 and response[1] != self.token: 
            raise PostmanError("Response token does not match request token.")
        self.token = (self.token + 1) & 0x7F
        return response[0:1] + response[2:]

    def put(self, path, data=None):
        self.send(TPM_PUT, self.token, path, data is not None, data)
        response = self.receive()
        if len(response) > 1 and response[1] != self.token: 
            raise PostmanError("Response token does not match request token.")
        self.token = (self.token + 1) & 0x7F
        return response[0:1] + response[2:]

    def post(self, path, data=None):
        self.send(TPM_POST, self.token, path, data is not None, data)
        response = self.receive()
        if len(response) > 1 and response[1] != self.token: 
            raise PostmanError("Response token does not match request token.")
        self.token = (self.token + 1) & 0x7F
        return response[0:1] + response[2:]

    def delete(self, path, query=None):
        self.send(TPM_DELETE, self.token, path, query is not None, query)
        response = self.receive()
        if len(response) > 1 and response[1] != self.token: 
            raise PostmanError("Response token does not match request token.")
        self.token = (self.token + 1) & 0x7F
        return response[0:1] + response[2:]
